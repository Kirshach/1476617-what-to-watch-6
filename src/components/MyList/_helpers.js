import React from 'react';

import MovieCardList from '../MovieCardList/MovieCardList';
import LoadingPlaceholder from '../LoadingPlaceholder/LoadingPlaceholder';

export const getFavouriteMoviesListComponent = (
    favouriteFilms,
    favouriteFilmsHaveLoaded,
) => {
  if (favouriteFilmsHaveLoaded) {
    return favouriteFilms.length > 0
      ? <MovieCardList films={favouriteFilms} />
      : (
        <div className="no-films-placeholder">
          There seems to be no sign of life in your favourite movies list
        </div>
      );
  } else {
    return <LoadingPlaceholder />;
  }
};
