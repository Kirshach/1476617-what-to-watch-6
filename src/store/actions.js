export const ActionType = {
  SET_GENRE: `films/setGenre`,
};

export const setGenre = (payload) => ({
  type: ActionType.SET_GENRE,
  payload
});
