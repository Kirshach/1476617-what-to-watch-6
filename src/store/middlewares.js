import {history} from '../history';

export const redirectActionType = `app/redirect`;

export const redirectAction = (payload) => ({
  type: redirectActionType,
  payload,
});

export const redirectMiddleware = (_store) => (next) => (action) => {
  if (action.type === redirectActionType) {
    history.push(action.payload);
  }
  return next(action);
};
