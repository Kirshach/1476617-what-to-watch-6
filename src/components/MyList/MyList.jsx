import React from 'react';

import PageFooter from '../PageFooter/PageFooter';
import PageHeader from '../PageHeader/PageHeader';

import {useFavouriteFilms} from '../../hooks';

import {getFavouriteMoviesListComponent} from './_helpers';

const MyList = () => {
  const [favouriteFilms, favouriteFilmsHaveLoaded] = useFavouriteFilms();

  return (
    <div className="user-page">
      <PageHeader className="user-page__head">
        <h1 className="page-title user-page__title">
          My list
        </h1>
      </PageHeader>

      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>

        <div className="catalog__movies-list">
          {getFavouriteMoviesListComponent(favouriteFilms, favouriteFilmsHaveLoaded)}
        </div>

      </section>

      <PageFooter />
    </div>
  );
};

MyList.propTypes = {};

export default MyList;
