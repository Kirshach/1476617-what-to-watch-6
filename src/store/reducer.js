import {ActionType} from './actions';
import {Genres} from '../const';
import {films} from '../mocks/films';


const initialState = {
  films,
  genre: Genres.allGenres,
  genres: Object.values(Genres),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
