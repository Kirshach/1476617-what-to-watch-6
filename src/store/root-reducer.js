import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';
import {domainReducer} from './domain/reducer';

export const DOMAIN = `domain`;
export const APP = `app`;

const rootReducer = combineReducers({
  [DOMAIN]: domainReducer,
  [APP]: appReducer,
});

export {rootReducer};

export const RootNS = {DOMAIN, APP};
