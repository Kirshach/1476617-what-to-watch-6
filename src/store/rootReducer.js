import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';
import {domainReducer} from './domain/reducer';

const DOMAIN = `domain`;
const APP = `app`;

const rootReducer = combineReducers({
  [DOMAIN]: domainReducer,
  [APP]: appReducer,
});

export {rootReducer};

export const RootNS = {DOMAIN, APP};
