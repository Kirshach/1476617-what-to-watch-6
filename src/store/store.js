import {createStore, applyMiddleware} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import {redirectMiddleware, handleAPIErrorMiddleware} from './middlewares';
import {checkAuthThunk} from '../store/app/auth/thunks';
import {fetchFilmsThunk} from '../store/domain/thunks';
import {rootReducer as reducer} from '../store/rootReducer';

import {api} from '../api/api';

const store = createStore(
    reducer,
    composeWithDevTools(
        applyMiddleware(
            handleAPIErrorMiddleware,
            redirectMiddleware,
            thunk.withExtraArgument(api),
        ),
    ),
);

store.dispatch(checkAuthThunk());
store.dispatch(fetchFilmsThunk());

export {store};
