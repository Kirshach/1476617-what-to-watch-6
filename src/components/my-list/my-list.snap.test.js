import React from 'react';
import {render} from '@testing-library/react';
import {MemoryRouter} from 'react-router';
import {useDispatch, useSelector} from 'react-redux';

import {store} from '../../mocks/store/store';

import MyList from './my-list';

function MockVideoPlayer() {
  return <div>This is a VideoPlayer placeholder</div>;
}

jest.mock(`react-redux`);
jest.mock(`../video-player/video-player`, () => ({
  __esModule: true,
  default: MockVideoPlayer,
}));

test(`<MyList /> renders correctly`, () => {
  useSelector.mockImplementation((selector) => selector(store));
  useDispatch.mockReturnValue(() => Promise.resolve());

  const {container} = render(
      <MemoryRouter>
        <MyList />
      </MemoryRouter>
  );
  expect(container).toMatchSnapshot();
});
