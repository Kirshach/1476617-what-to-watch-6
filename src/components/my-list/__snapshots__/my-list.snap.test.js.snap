// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyList /> renders correctly 1`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        My list
      </h1>
      <div
        class="user-block"
      >
        <a
          class="user-block__link"
          href="/mylist"
        >
          <div
            class="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg"
              width="63"
            />
          </div>
        </a>
      </div>
    </header>
    <section
      class="catalog"
    >
      <h2
        class="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <div
        class="catalog__movies-list"
      >
        <article
          class="small-movie-card catalog__movies-card"
          style="cursor: pointer;"
        >
          <div
            class="small-movie-card__image"
          >
            <div>
              This is a VideoPlayer placeholder
            </div>
          </div>
          <h3
            class="small-movie-card__title"
            style="pointer-events: none;"
          >
            <span
              class="small-movie-card__link"
            >
              Macbeth
            </span>
          </h3>
        </article>
        <article
          class="small-movie-card catalog__movies-card"
          style="cursor: pointer;"
        >
          <div
            class="small-movie-card__image"
          >
            <div>
              This is a VideoPlayer placeholder
            </div>
          </div>
          <h3
            class="small-movie-card__title"
            style="pointer-events: none;"
          >
            <span
              class="small-movie-card__link"
            >
              Beach
            </span>
          </h3>
        </article>
      </div>
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
